### PR TITLE
Add filters.

### PR DESCRIPTION
**Pull request by floitsch**
Instead of strings we now have a `Filter` class and many top-level convenience methods to create them.

With this patch it is now possible to use more operators and also to create logical combination ('or', 'and') which was previously not possible. Similarly, it was not possible to test for 'null' values.

### CODE REVIEW

File: src/filter.toit [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+
+import .supabase
+
+/**
+Provides ways to create $Filter objects for
+  $PostgRest.select, $PostgRest.delete and $PostgRest.update
+  operations.
+
+Functions in this library create $Filter objects that can be used
+  to filter the rows that are returned by the database.
```
Comment (erikcorry) [inline src/filter.toit]: Filters on the server side?

File: src/filter.toit [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+
+Most of the functionality can be accessed through convenience functions,
+  such as $equals, $less_than, or similar. More complex functionality
+  might need direct calls to the $Filter constructors. Specifically,
+  $Filter.raw allows to create any filter.
```
Comment (erikcorry) [inline src/filter.toit]:
```suggestion
  $Filter.raw allows one to create any filter.
```

File: src/filter.toit [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+
+  constructor .column .values:
+
+  to_string --nested/bool --negated/bool -> string:
+    not_string := negated ? "not." : ""
+    column_string := Filter.encode_column column
+    encoded_values := values.map: Filter.encode it --may_quote
+    joined := encoded_values.join ","
+    return "$column_string$(nested ? "." : "=")$not_string$Filter.IN.($joined)"
+
```
Comment (erikcorry) [inline src/filter.toit:479]: Extra blank line.

File: src/filter.toit [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+    column_string := Filter.encode_column column
+    encoded_values := values.map: Filter.encode it --may_quote
+    joined := encoded_values.join ","
+    return "$column_string$(nested ? "." : "=")$not_string$op.{$joined}"
+
+class FilterRange_ implements Filter:
+  column/string
+  op/string
+  from/any  // Can be a num, or a date.
+  to/any    // Can be a num, or a date.
```
Comment (erikcorry) [inline src/filter.toit:498]:
```suggestion
  from/any  // Can be a num, or a Date.
  to/any    // Can be a num, or a Date.
```

File: src/filter.toit [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+    return filter.to_string --nested=nested --negated=(not negated)
+
+class FilterRaw_ implements Filter:
+  raw/string
+
+  constructor .raw:
+
+  to_string --nested/bool --negated/bool -> string:
+    if nested or negated:
+      throw "RAW MUST NOT BE NESTED"
```
Comment (erikcorry) [inline src/filter.toit]:
```suggestion
      throw "RAW_MUST_NOT_BE_NESTED"
```

File: tests/supabase_filters_test.toit [hunk 6]
```diff
@@ -0,0 +1,373 @@
+// Copyright (C) 2023 Toitware ApS.
```
Comment (erikcorry) [inline tests/supabase_filters_test.toit:1]: No message about the license file.

**Review by erikcorry — APPROVED** [review]

File: src/filter.toit (hunk 1 above)
Comment (floitsch) [reply src/filter.toit]: done.

File: src/filter.toit (hunk 2 above)
Comment (floitsch) [reply src/filter.toit]: done.

File: src/filter.toit (hunk 3 above)
Comment (floitsch) [reply src/filter.toit:479]: done.

File: src/filter.toit (hunk 4 above)
Comment (floitsch) [reply src/filter.toit:498]: Leaving as is, as dates in strings are also allowed.

File: src/filter.toit (hunk 5 above)
Comment (floitsch) [reply src/filter.toit]: done.

File: tests/supabase_filters_test.toit (hunk 6 above)
Comment (floitsch) [reply tests/supabase_filters_test.toit:1]: Will update the license in all the test files in a separate PR.

**Pull request merged**